openfermioncirq/gates/fermionic_simulation.py

```diff
@@ -86,11 +87,57 @@ def state_swap_eigen_component(x: str, y: str, sign: int = 1, angle: float=0):
     return component
 
 
+@cirq.value_equality(approximate=True)
+class FermionicSimulationGate(cirq.EigenGate):
+    """The Jordan-Wigner transform of exp(-i H) for a fermionic Hamiltonian
+    H."""
+
+    def __init__(self,
+                 weights: Optional[Tuple[complex, ...]]=None,
+                 absorb_exponent: bool = False,
+                 **kwargs) -> None:
+        if weights is None:
+            weights = (1.,) * self.num_weights
+        self.weights = weights
+
+        super().__init__(**kwargs)
+
+        if absorb_exponent:
+            self.absorb_exponent_into_weights()
+
+    @abc.abstractproperty
+    def num_weights(self) -> int:
+        """The number of parameters (weights) in the generator."""
```

Comment:
The description of this class says that it represents `exp(-iH)`, but it doesn't contain enough information to identify what `H` actually is. I think this information needs to be present. This class should store not only the number of weights, but also what generators the weights correspond to. Perhaps the EigenGate methods can then be implemented here? If not, then I don't think we should inherit from EigenGate here.

Reply:
The mapping from weights to Hamiltonian/generator is specific to the subclass (`QuadraticFermionicSimulationGate`, etc.), and this information is in the docstring for each of them. I could add this to the docstring for `FermionicSimulationGate`.

Having `FermionicSimulationGate` (rather than its subclasses) inherit `EigenGate` is nice for type checking.

Reply:
Subclasses should be forced to specify the mapping from weights to generator. So this class should have an abstract method to that effect. I'm thinking something like
```
def generators(self) -> Tuple[FermionOperator]:
```
The number of weights can be inferred from this. Currently the only abstract method you have, `num_weights`, is not enough to actually specify what gate is being represented. I think that is a serious flaw.

I still think that we should either implement EigenGate here (it might be possible once you force specifying the generators), or not inherit it at all. There's no reason to force subclasses to implement an unnecessary interface.

Reply:
I think I understand. #353 has generators, but in the qubit representation. I could definitely add the fermionic generator as well. (This is already in the docstring, but it would be nice to have it accessible programmatically.) For efficiency though, it's probably best to 
keep the qubit generator (with tests for consistency), in which case `_eigen_components` would still be implemented by the subclasses. Otherwise, `FermionicSimulationGate` needs to do an unnecessary JW transformation.

The argument for having `FermionicSimulationGate` inherit from `EigenGate` is that it forces all subclasses to implement `_eigen_compoments` and tells mypy as much. What's the argument against inheriting `EigenGate`, even if `_eigen_components` is implemented by the subclasses?

Reply:
Also, could the fermionic generator thing wait for a later PR? Makes sense to add it together with the qubit generator.

Reply:
For this PR you should store the generator. Since the name of this class contains "Fermion", the stored generator should be a FermionOperator or other fermionic class. If you want to implement EigenGate here, then you'll need to perform the JW transformation. Subclasses can still override for efficiency.

The argument against inheriting EigenGate without actually implementing it is that it causes extra work for somebody who wants to create a FermionicSimulationGate but doesn't need EigenGate functionality. I still don't understand why it is advantageous for mypy to think that this is an EigenGate.